Lightweight Storefront: Theme Picker UI

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11326
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This adds the SwiftUI view for the theme carousel, as well as the SwiftUI view for the Profiler theme picker (which makes use of the aforementioned carousel).

It uses example URLs to show the screenshots and should be replaced with real data.

- ThemeCarouselView` is meant to be a reusable View as there's two places that will use it, Profiler and Settings.
- `ProfilerThemesPickerView` is meant to show how to make use of `ThemeCarouselView`


## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

They're not used in the app yet, so test by previewing `ThemesCarouselView` and `ProfilerThemesPickerView`. Especially please test the scrolling behavior and make sure the last message is also shown.


| Carousel default view | Carousel missing image | Carousel last message |
|--------|--------|--------|
| <img width="342" alt="Screenshot 2023-12-01 at 13 15 59" src="https://github.com/woocommerce/woocommerce-ios/assets/266376/e3d789bc-36a6-44ed-bd0e-a97b835a7e7e"> | <img width="337" alt="Screenshot 2023-12-01 at 13 16 12" src="https://github.com/woocommerce/woocommerce-ios/assets/266376/4f36567d-4dee-42e1-a8bf-3705e4f56ed8"> | <img width="335" alt="Screenshot 2023-12-01 at 13 16 18" src="https://github.com/woocommerce/woocommerce-ios/assets/266376/239e8285-a409-4361-91f8-412f65ddb9a7">  |

`ProfilerThemesPickerView`:

<img width="365" alt="Screenshot 2023-12-01 at 13 16 33" src="https://github.com/woocommerce/woocommerce-ios/assets/266376/8221e10f-3d14-4b6d-b213-95b1994262ea">

hafizrahman committed:
Add starter file for Themes picker.

hafizrahman committed:
Add carousel that loads image from URL.

hafizrahman committed:
Add last message at the end of carousel, and some tweaks.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPickerView.swift:
```diff
@@ -0,0 +1,149 @@
+    @ScaledMetric private var scale: CGFloat = 1.0
+
+    var body: some View {
+        let urls = [
```
Just example URLs for testing purposes.

hafizrahman committed:
Update hardcoded values.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPickerView.swift:
```diff
@@ -0,0 +1,149 @@
+                    // todo
+                }
+            }
+        }
```
not showing yet on the design for now.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPickerView.swift:
```diff
@@ -0,0 +1,150 @@
+                    // TODO: Setup toolbar.
+                }
+            }
+        }
```
Not showing anything yet for now.

hafizrahman committed:
Move carousel into its own View to make it reusable on different places.

hafizrahman committed:
Add ProfilerThemesPickerView.

hafizrahman committed:
Delete unused file.

hafizrahman committed:
Fix trailing line.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+            Text(name)
+        }
+        .frame(width: Layout.imageWidth, height: Layout.imageHeight)
+        .background(Color.withColorStudio(
+            name: .wooCommercePurple,
+            shade: .shade0))
```
Not sure about this color, just picking a nice color since there is no design for it at the moment.

itsmeichigo replied:
Yup, please check with Ann for the appropriate color.

hafizrahman replied:
Asked in figma

hafizrahman commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+
+struct ThemesCarouselView: View {
+    /// Tuple array containing theme name and theme Url string pair.
+    let themesInfo: [(name: String, url: String)] = [
+        (name: "Google", url: "google.com"),
+        (name: "Facebook", url: "facebook.com"),
+        (name: "WordPress", url: "wordpress.com"),
+        (name: "Microsoft", url: "microsoft.com"),
+        (name: "Amazon", url: "amazon.com"),
+        (name: "Apple", url: "apple.com")
+    ]
```
default values just for demo purposes and in practice they'll have to be modified to accept actual data.

hafizrahman commented:
@itsmeichigo This is now ready to review. 

Since we're working in parallel, just want to mention that this PR is my idea how the Views usage should be. It can be subject to some changes depending on how you think things should be, so happy to discuss and refactor as necessary.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ProfilerThemesPickerView.swift:
```diff
@@ -0,0 +1,93 @@
+
+        ScrollView {
+            VStack(alignment: .leading) {
+                Spacer()
+                    .frame(height: Layout.contentVerticalSpacing)
+
+                Text(Localization.chooseThemeHeading)
+                    .bold()
+                    .largeTitleStyle()
+                    .padding(.horizontal, Layout.contentPadding)
+
+                Spacer()
+                    .frame(height: Layout.contentPadding)
```
Instead of adding spacer with fixed height, we can simply add vertical paddings to the `Text`:

```suggestion
                Text(Localization.chooseThemeHeading)
                    .bold()
                    .largeTitleStyle()
                    .padding(.horizontal, Layout.contentPadding)
                    .padding(.top, Layout.contentVerticalSpacing)
                    .padding(.bottom, Layout.contentPadding)
```

hafizrahman replied:
Nice! Much more shorter for the same effect.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ProfilerThemesPickerView.swift:
```diff
@@ -0,0 +1,93 @@
+                    lastMessageContent: Localization.lastMessageContent
+                )
+
+                Spacer()
```
Since we want the carousel to be in the center of the remaining area, how about adding just flexible sizes for the `Spacer`'s:

```suggestion
                Spacer()

                ThemesCarouselView(
                    lastMessageHeading: Localization.lastMessageHeading,
                    lastMessageContent: Localization.lastMessageContent
                )

                Spacer()
```

hafizrahman replied:
Another case where I followed the design too literally. This looks better! 👍🏼

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ProfilerThemesPickerView.swift:
```diff
@@ -0,0 +1,93 @@
+import SwiftUI
+import Kingfisher
+
+struct ProfilerThemesPickerView: View {
```
Let's keep the consistent pattern of adding comments for each new type explaining what it's for.

```suggestion
/// View for picking themes in the store creation flow.
struct ProfilerThemesPickerView: View {
```

hafizrahman replied:
Oh thanks! I forgot this when replacing the file.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ProfilerThemesPickerView.swift:
```diff
@@ -0,0 +1,93 @@
+
+    private enum Localization {
+        static let skipButtonTitle = NSLocalizedString(
+            "ThemesPickerView.skipButtonTitle",
```
Nit: we should follow the reverse domain name notation, which includes keeping the names in lower case (same for other keys below).

hafizrahman replied:
Ah, you meant lowercase the first letter, right? I missed that. Thanks!

hafizrahman replied:
https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/74b07159e053d3b75b111128957b710e56c2d58c

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+
+                // Demo for correct state
+                ForEach(themesInfo, id: \.name) { theme in
+                    if let themeUrl = getThemeUrl(themeUrl: theme.url) {
+                        ThemeImage(url: themeUrl)
+                    }
```
Should we also fall back to `ThemeName` if the url cannot be found?

```suggestion
                    if let themeUrl = getThemeUrl(themeUrl: theme.url) {
                        ThemeImage(url: themeUrl)
                    } else {
                        ThemeName(name: theme.name)
                    }
```

hafizrahman replied:
Ah yes, I intentionally didn't add error handling here (same as there's no error handling in `ThemeImage` yet if the image can't be loaded) as this is just UI and I'm thinking to add them a separate PR. But now that I think of it, this is straightforward enough to add now.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+                ThemeName(name: themesInfo[0].name)
+
+                // Demo for last message content
+                VStack {
+                    Text(lastMessageHeading)
+                        .bold()
+                        .bodyStyle()
+                        .padding(.horizontal, Layout.contentPadding)
+
+                    Spacer()
+                        .frame(height: Layout.contentPadding)
+
+                    Text(lastMessageContent)
+                        .subheadlineStyle()
+                            .multilineTextAlignment(.center)
+                        .padding(.horizontal, Layout.contentPadding)
+                }.frame(
```
Instead of adding spacer with fixed height, we should add spacing to the outer `VStack` to simplify the layout:

```suggestion
                VStack(spacing: Layout.contentPadding) {
                    Text(lastMessageHeading)
                        .bold()
                        .bodyStyle()
                        .padding(.horizontal, Layout.contentPadding)

                    Text(lastMessageContent)
                        .subheadlineStyle()
                            .multilineTextAlignment(.center)
                        .padding(.horizontal, Layout.contentPadding)
                }.frame(
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+private struct ThemeImage: View {
```
Nit: How about naming this to `ThemeThumbnailCard`?

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+        KFImage(url)
+            .resizable()
+            .aspectRatio(contentMode: .fit)
+            .frame(width: Layout.imageWidth * scale, height: Layout.imageHeight * scale)
```
Adding scale to the images may make it too big to see when using a large font size. Maybe it's better to keep the size fixed?

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+                leading: Layout.imageHorizontalPadding,
+                bottom: Layout.imageVerticalPadding,
+                trailing: Layout.imageHorizontalPadding)
+            )
```
I see that `imageHorizontalPadding` and `imageVerticalPadding` have the same value, so we can simplify this to:

```Swift
.padding(Layout.imageHorizontalPadding)
```

If you have different sizes for each edge, consider adding a `UIEdgeInset` in the `Layout` enum and use it here for simplicity.

hafizrahman replied:
https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/021cf3e3d26eb85fe5944edc2b5e6bd0946105a6

I had these paddings added separately when debugging why the shadow was cut off at the top and bottom, and forgot to remove them once it's solved. Thanks for the reminder.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+            shade: .shade0))
+        .cornerRadius(Layout.cornerRadius)
+        .shadow(radius: Layout.shadowRadius, x: 0, y: Layout.shadowYOffset)
+        .padding(.init(
+            top: Layout.imageVerticalPadding,
+            leading: Layout.imageHorizontalPadding,
+            bottom: Layout.imageVerticalPadding,
+            trailing: Layout.imageHorizontalPadding)
+        )
```
Same comment as above.

hafizrahman replied:
https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/021cf3e3d26eb85fe5944edc2b5e6bd0946105a6

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+private struct ThemeName: View {
```
Nit: How about renaming this to `ThemeNameCard`?

itsmeichigo replied:
For simplicity, you can also create a method in a private extension of `ThemesCarouselView` instead of adding a new type:

```Swift
func themeNameCard(name: String) -> some View {
    VStack {
        #content-here#
    }
}
```

hafizrahman replied:
Nice! I wasn't sure when to make a new type and when to use func, but I looked it up and this makes sense. I refactored both the image and name card on https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/64996abe04324b2be7ec7ebf70f7801143f7a98a

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+}
+
+private extension ThemesCarouselView {
+    private func getThemeUrl(themeUrl: String) -> URL? {
```
Nit: How about naming this `getThemeThumbnailURL` to clarify the intention?

itsmeichigo replied:
I'd suggest adding an extension method for `WordPressTheme` to get this detail as it looks to be fixed:

```Swift
extension WordPressTheme {
    var themeThumbnailURL: URL? {
         // compute the URL and return it
    }
}
```

hafizrahman replied:
I added the extension method in https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/9a7b191244786cd0a575fa8f8ec51ff2e3d34d42

It's not currently in use, though (since the UI still uses hardcoded value instead of actually using `WordPressTheme`), but I've left a todo comment to remind us to use the real method and delete that `getThemeUrl` function.

itsmeichigo reviewed (commented):
This looks nice! I like how reusable the carousel view is, with the option to change the text at the end 💯 

It would be even better if the scroll view can snap into each item. I found this [article](https://www.hackingwithswift.com/quick-start/swiftui/how-to-make-a-scrollview-snap-with-paging-or-between-child-views), which introduced a new API allowing this from iOS 17. I'd suggest moving the carousel scroll view to an extension method and apply the new modifier for the scroll view with the `#available(iOS 17, *)` check.

Besides, I also checked the view for accessibility and looks like the scale makes the theme thumbnail a bit too big to see. I'd suggest removing the use of scale and maybe make the content within the last slide scrollable just in case.


https://github.com/woocommerce/woocommerce-ios/assets/5533851/25155e07-657d-4cea-9235-6ee8fd927dda

I also left some suggestions in comments below.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+                VStack {
+                    Text(lastMessageHeading)
+                        .bold()
+                        .bodyStyle()
```
Nit: the text in the design seems to have secondary color.

<img width="245" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/cc445b20-a724-4c00-befc-98745f4bb831">

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/b957e7b29b97691e65b15f9c5b98ba0ce42e9022

hafizrahman committed:
Add comment explaining the view.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Fix reverse domain notation naming.

hafizrahman committed:
Handle missing theme url case.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Simplify Text spacing.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Update spacing around ThemeCarouselView

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Simplify paddings.

hafizrahman committed:
Refactor sub views to be func instead, and remove scaling for image.

hafizrahman committed:
Refactor the last message.

1. Make it scrollable on bigger font size while still matching the images' size.
2. Updates font color.

hafizrahman commented:
@itsmeichigo The theme thumbnail and last message sizing suggestions are handled in 

https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/b957e7b29b97691e65b15f9c5b98ba0ce42e9022

hafizrahman commented:
> It would be even better if the scroll view can snap into each item. I found this [article](https://www.hackingwithswift.com/quick-start/swiftui/how-to-make-a-scrollview-snap-with-paging-or-between-child-views), which introduced a new API allowing this from iOS 17. I'd suggest moving the carousel scroll view to an extension method and apply the new modifier for the scroll view with the #available(iOS 17, *) check.

This is so cool. I'll push it to a separate PR because for my work setup I still need to update macOS and Xcode to get it 🙏🏼

hafizrahman committed:
Merge branch 'trunk'

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+}
+
+private extension ThemesCarouselView {
+    func themeImage(url: URL) -> some View {
```
Nit: how about `themeImageCard` for consistency?

hafizrahman replied:
Yup! Added in https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/fb5975eda0c7bf0a0c2a54ff170667ff3c3482be

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesCarouselView.swift:
```diff
@@ -0,0 +1,130 @@
+                    VStack {
+                        Spacer()
+                        Text(lastMessageHeading)
+                            .foregroundColor(Color(.secondaryLabel))
+                            .bold()
+                            .bodyStyle()
```
Nit: this can be simplified with `secondaryBodyStyle`:

```suggestion
                            .bold()
                            . secondaryBodyStyle()
```

hafizrahman replied:
Added in https://github.com/woocommerce/woocommerce-ios/pull/11336/commits/0e8687a65622534c196a1edca98825b822a59fb3

itsmeichigo reviewed (approved):
Thanks for the update! I left some nitpicking comments but please feel free to merge this PR after addressing them.

hafizrahman committed:
Rename for consistency

hafizrahman committed:
Simplify styling.

hafizrahman committed:
Merge branch 'trunk'

hafizrahman committed:
Add extension variable for `WordPressTheme` to get the thumbnail URL.

Also explains in the preview that the current function is just temporary.

hafizrahman commented:
Thank you for your thorough review, @itsmeichigo , I learned a lot more about Swift UI!

I'm merging this so we can carry on with the actual functionality. There are two items that will need to be handled in future PRs:

1. The card color when the image can't be loaded. Currently the conversation is still undecided whether we want to keep the design or hide the card completely if the image can't be loaded (check Figma), so let's keep it this way for now and update later in a separate PR.
2. `private func getThemeUrl(themeUrl: String)` needs to be deleted once actual functionality making use of this View is being worked on.
